### PR TITLE
add target unit to retrieve latest

### PR DESCRIPTION
**Pull request by matiascognite**
## Description
- Add target_unit and target_unit_system to retrieve_latest

## Checklist:
- [x] Tests added/updated.
- [x] Documentation updated. Documentation is generated from docstrings - these must be updated according to your change.
  If a new method has been added it should be referenced in [cognite.rst](https://github.com/cognitedata/cognite-sdk-python/blob/master/docs/source/cognite.rst) in order to generate docs based on its docstring.
- [x] Changelog updated in [CHANGELOG.md](https://github.com/cognitedata/cognite-sdk-python/blob/master/CHANGELOG.md).
- [x] Version bumped. If triggering a new release is desired, bump the version number in [_version.py](https://github.com/cognitedata/cognite-sdk-python/blob/master/cognite/client/_version.py) and [pyproject.toml](https://github.com/cognitedata/cognite-sdk-python/blob/master/pyproject.toml) per [semantic versioning](https://semver.org/).

### CODE REVIEW

**Comment by matiascognite** [comment]
maybe beta removal and target unit should be two separate PRs, if so, I can split them up

**Comment by matiascognite** [comment]
Not sure why the tests fail, the failures seem unrelated to my change?

File: cognite/client/_api/datapoints.py
Comment (haakonvt) [inline cognite/client/_api/datapoints.py:1]: `LatestDatapointQuery` should accept the unit arguments to allow customization similar to how `before` is treated.

File: tests/tests_integration/test_api/test_datapoints.py [hunk 1]
```diff
@@ -1793,6 +1793,28 @@ def test_retrieve_latest_before(self, cognite_client, all_test_time_series):
...
+        res = cognite_client.time_series.data.retrieve_latest(
+            external_id=timeseries.external_id, before="now", **kwargs
+        )
+
+        if isinstance(res, pd.DataFrame):
+            res = DatapointsArray(value=res.values)
+
+        assert math.isclose(res.value[0], 212)
+        assert res.unit_external_id == "temperature:deg_f"
+
```
Comment (haakonvt) [inline tests/tests_integration/test_api/test_datapoints.py:1817]: Perhaps a test where you pass both unit args?

**Comment by haakonvt** [comment]
> maybe beta removal and target unit should be two separate PRs, if so, I can split them up

That would be a good idea, as there is more to removing the warning that this PR currently does: e.g. updates to time series objects (whether through update or upsert method) depend on whether units is present or not (i.e. an upsert with replace mode won't clear unit external id unless we decide we need to use the beta endpoint).

**Comment by matiascognite** [comment]
@haakonvt Thanks for the review! I've tried to modify the PR accordingly (adding target to LatestDatapointQuery, adding tests, don't remove beta).

Still need to create changelog and bump version, and still don't know why the (unrelated?) tests fail.

File: tests/tests_integration/test_api/test_datapoints.py [hunk 2]
```diff
@@ -1793,6 +1793,63 @@ def test_retrieve_latest_before(self, cognite_client, all_test_time_series):
...
+        timeseries_degree_c_minus40_0_100: TimeSeries,
+    ) -> None:
+        timeseries = timeseries_degree_c_minus40_0_100
+
+        res = cognite_client.time_series.data.retrieve_latest(
+            external_id=timeseries.external_id, before="now", **kwargs
+        )
+
+        if isinstance(res, pd.DataFrame):
+            res = DatapointsArray(value=res.values)
```
Comment (haakonvt) [inline tests/tests_integration/test_api/test_datapoints.py:1813]: Don't see how a data frame can ever be returned?
```suggestion
```

File: tests/tests_integration/test_api/test_datapoints.py [hunk 3]
```diff
@@ -1793,6 +1793,63 @@ def test_retrieve_latest_before(self, cognite_client, all_test_time_series):
...
+        timeseries_degree_c_minus40_0_100: TimeSeries,
+    ) -> None:
+        timeseries = timeseries_degree_c_minus40_0_100
+
+        res = cognite_client.time_series.data.retrieve_latest(
+            external_id=LatestDatapointQuery(external_id=timeseries.external_id, before="now", **kwargs)
+        )
+
+        if isinstance(res, pd.DataFrame):
+            res = DatapointsArray(value=res.values)
```
Comment (haakonvt) [inline tests/tests_integration/test_api/test_datapoints.py:1835]: Same here:
```suggestion
```

File: cognite/client/_api/datapoints.py [hunk 4]
```diff
@@ -1189,6 +1191,8 @@ def retrieve_latest(
             id (int | LatestDatapointQuery | list[int | LatestDatapointQuery] | None): Id or list of ids.
             external_id (str | LatestDatapointQuery | list[str | LatestDatapointQuery] | None): External id or list of external ids.
             before (None | int | str | datetime): (Union[int, str, datetime]): Get latest datapoint before this time. Not used when passing 'LatestDatapointQuery'.
+            target_unit (str | None): The unit_external_id of the data points returned. If the time series does not have a unit_external_id that can be converted to the target_unit, an error will be returned. Cannot be used with target_unit_system.
```
Comment (haakonvt) [inline cognite/client/_api/datapoints.py:1194]:
```suggestion
            target_unit (str | None): The unit_external_id of the data point returned. If the time series does not have a unit_external_id that can be converted to the target_unit, an error will be returned. Cannot be used with target_unit_system.
```

File: cognite/client/_api/datapoints.py [hunk 5]
```diff
@@ -1189,6 +1191,8 @@ def retrieve_latest(
...
+            target_unit_system (str | None): The unit system of the data points returned. Cannot be used with target_unit.
```
Comment (haakonvt) [inline cognite/client/_api/datapoints.py:1195]:
```suggestion
            target_unit_system (str | None): The unit system of the data point returned. Cannot be used with target_unit.
```

File: cognite/client/_api/datapoints.py [hunk 6]
```diff
@@ -1231,7 +1235,9 @@ def retrieve_latest(
                 ...     id=id_queries,
                 ...     external_id=LatestDatapointQuery(external_id="abc", before="3h-ago"))
         """
-        fetcher = RetrieveLatestDpsFetcher(id, external_id, before, ignore_unknown_ids, self)
+        fetcher = RetrieveLatestDpsFetcher(
```
Comment (haakonvt) [inline cognite/client/_api/datapoints.py:1238]: Let's add an example to the docstring showcasing how to specify unit 🚀

**Review by haakonvt — COMMENTED** [review]
Nice! Final round of comments from me 😄

File: cognite/client/_api/datapoints.py [hunk 7]
```diff
@@ -1634,6 +1650,16 @@ def _prepare_requests(
                 i_before = self.before_settings.get((identifier_type, i), self.default_before)
                 if "now" != i_before is not None:  # mypy doesn't understand 'i_before not in {"now", None}'
                     dct["before"] = timestamp_to_ms(i_before)
+                i_target_unit = self.target_unit_settings.get((identifier_type, i), self.default_unit)
+                i_target_unit_system = self.target_unit_system_settings.get(
+                    (identifier_type, i), self.default_unit_system
+                )
```
[the remark this answers is not in the document]
Comment (haakonvt) [reply cognite/client/_api/datapoints.py:1656]: When I think about it, for the `before` arg, you can't use the `or` notation as `before=0` is valid and should not be changed to the default 😅

File: cognite/client/_api/datapoints.py (hunk 6 above)
Comment (matiascognite) [reply cognite/client/_api/datapoints.py:1243]: Done

File: cognite/client/_api/datapoints.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (matiascognite) [reply cognite/client/_api/datapoints.py]: had to modify it slightly, to avoid keyerror, hope it is correct now

**Review by haakonvt — APPROVED** [review]
# 🚀 🚀 🚀

**Pull request merged**